packages/graphql/src/classes/Subgraph.ts

```diff
@@ -0,0 +1,291 @@
+        FederationDirectiveName | FullyQualifiedFederationDirectiveName | string
+    >;
+    private typeDefs: TypeSource;
+    private linkExtension: SchemaExtensionNode;
+
+    constructor(typeDefs: TypeSource) {
+        this.typeDefs = typeDefs;
+
+        this.importArgument = new Map([
+            ["key", "federation__key"],
+            ["extends", "federation__extends"],
+            ["shareable", "federation__shareable"],
+            ["inaccessible", "federation__inaccessible"],
+            ["override", "federation__override"],
+            ["external", "federation__external"],
+            ["provides", "federation__provides"],
+            ["requires", "federation__requires"],
+            ["tag", "federation__tag"],
+        ]);
+
+        const linkMeta = this.findFederationLinkMeta(typeDefs);
+        if (!linkMeta) {
+            throw new Error(`typeDefs must contain \`@link\` schema extension to be used with Apollo Federation`);
+        }
+        const { extension, directive: linkDirective } = linkMeta;
+
+        this.linkExtension = extension;
+
+        this.parseLinkImportArgument(linkDirective);
+    }
+
+    public getFullyQualifiedDirectiveName(
+        name: FederationDirectiveName
+    ): FederationDirectiveName | FullyQualifiedFederationDirectiveName | string {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return this.importArgument.get(name)!;
+    }
+
+    // TODO: Remove this function, parse `@shareable` directives and add them only on related generated types and fields
+    public augmentGeneratedSchemaDefinition(typeDefs: DocumentNode): DocumentNode {
+        const definitions: DefinitionNode[] = [];
+
+        const shareable = makeDirectiveNode(this.importArgument.get("shareable") as string, {}) as ConstDirectiveNode;
+
+        // This is a really filthy hack to apply @shareable
+        for (const definition of typeDefs.definitions) {
+            if (definition.kind === Kind.OBJECT_TYPE_DEFINITION) {
+                if (
+                    ["Query", "Mutation"].includes(definition.name.value) ||
+                    definition.name.value.endsWith("Connection") ||
+                    definition.name.value.endsWith("AggregateSelection") ||
+                    definition.name.value.endsWith("Edge")
+                ) {
+                    if (definition.directives) {
+                        definitions.push({
+                            ...definition,
+                            directives: [...definition.directives, shareable],
+                        });
+                    } else {
+                        definitions.push({
+                            ...definition,
+                            directives: [shareable],
+                        });
+                    }
+                } else {
+                    definitions.push(definition);
+                }
+            } else {
+                definitions.push(definition);
+            }
```

Comment:
Could this be slightly more readable if it where in a separate function? 

Maybe we could get rid of this ultra-nested `else`(s) with a function like:

```ts
private parseDefinition(definition): DefinitionNode
```